crates/order-validation/src/banned.rs

```diff
@@ -53,28 +53,60 @@ impl Users {
         }
     }
 
-    /// Returns `true` if the given address is banned or an error if the result
-    /// cannot be determined.
-    pub async fn is_banned(&self, address: H160) -> Result<bool, Error> {
-        if self.list.contains(&address) {
-            return Ok(true);
-        }
+    /// Returns a subset of addresses from the input iterator which are banned.
+    pub async fn banned(&self, addresses: impl Iterator<Item = H160>) -> HashSet<H160> {
+        let mut banned = HashSet::new();
+
+        let need_lookup = addresses
+            .filter(|address| {
+                if self.list.contains(address) {
+                    banned.insert(*address);
+                    false
+                } else {
+                    true
+                }
+            })
+            // Need to collect here to make sure filter gets executed and we insert addresses
+            .collect::<HashSet<_>>();
 
         if let Some(onchain) = &self.onchain {
```

Comment:
Since this function is getting quite big I think removing 1 level of indentation with an early return might be good.
```suggestion
        let Some(onchain) = &self.onchain else { return banned; }
```